File: bindings/shape.cpp
#include <pybind11/pybind11.h>

#include "binding_utils/utils.hpp"

#include <stratax/core/Shape.hpp>
#include <stratax/exceptions/Exceptions.hpp>

#include <cstddef>
#include <sstream>
#include <vector>

namespace py = pybind11;

namespace
{

using Shape = stratax::core::Shape;

long long cast_shape_dimension(py::handle dim)
{
    return binding_utils::cast_integer(dim);
}

Shape make_shape(const std::vector<long long>& dims)
{
    std::vector<std::size_t> values;
    values.reserve(dims.size());

    for (long long dim : dims)
    {
        if (dim < 0)
        {
            throw Exceptions::ShapeError("Shape dimensions cannot be negative.");
        }

        values.push_back(binding_utils::checked_size(dim));
    }

    return Shape(values);
}

} // anonymous namespace

void bind_shape(py::module_& m)
{
    py::class_<Shape> cls(m, "Shape");

    cls
        .def(py::init<>())
        .def(py::init<const Shape&>(), py::arg("other"))
        .def(
            py::init([](py::int_ size) {
                return make_shape({cast_shape_dimension(size)});
            }),
            py::arg("size"))
        .def(
            py::init([](py::iterable dims) {
                std::vector<long long> values;

                for (py::handle dim : dims)
                {
                    values.push_back(cast_shape_dimension(dim));
                }

                return make_shape(values);
            }),
            py::arg("dims"))
        .def_property_readonly("rank", &Shape::rank)
        .def_property_readonly(
            "elements",
            [](const Shape& shape) {
                try
                {
                    return shape.elements();
                }
                catch (const Exceptions::DimensionError&)
                {
                    binding_utils::raise_overflow(
                        Exceptions::OverflowError("Shape element count overflow."));
                }
            })
        .def_property_readonly("empty", &Shape::empty)
        .def_property_readonly("strides", &Shape::strides)
        .def("__len__", &Shape::rank)
        .def("__getitem__", [](const Shape& shape, py::handle index) {
            return shape.at(binding_utils::cast_index(index));
        })
        .def(
            "__iter__",
            [](const Shape& shape) {
                return py::make_iterator(shape.begin(), shape.end());
            },
            py::keep_alive<0, 1>())
        .def("__eq__", [](const Shape& lhs, const Shape& rhs) {
            return lhs == rhs;
        })
        .def("__ne__", [](const Shape& lhs, const Shape& rhs) {
            return lhs != rhs;
        })
        .def("__repr__", [](const Shape& shape) {
            std::ostringstream os;
            os << shape;
            return os.str();
        });
}

File: python/stratax/shape.py
from __future__ import annotations

from collections.abc import Iterable

from ._core import _Shape
from .exceptions import TypeError as StrataxTypeError


class Shape:
    def __init__(self, *args):
        if not args:
            self._impl = _Shape()

        elif len(args) == 1:
            value = args[0]

            if (
                isinstance(value, _Shape)
                or not isinstance(value, Shape)
                and isinstance(value, Iterable)
            ):
                self._impl = _Shape(value)

            elif isinstance(value, Shape):
                self._impl = _Shape(value._impl)

            else:
                raise StrataxTypeError(
                    "Shape constructor expects a Shape or iterable of integers."
                )

        else:
            self._impl = _Shape(args)

    @property
    def rank(self):
        return self._impl.rank()

    @property
    def elements(self):
        return self._impl.elements()

    @property
    def empty(self):
        return self._impl.empty()

    def __len__(self):
        return len(self._impl)

    def __getitem__(self, index):
        return self._impl[index]

    def __iter__(self):
        return iter(self._impl)

    def __eq__(self, other):
        if not isinstance(other, Shape):
            return NotImplemented

        return self._impl == other._impl

    def __ne__(self, other):
        if not isinstance(other, Shape):
            return NotImplemented

        return self._impl != other._impl

    def __repr__(self):
        return repr(self._impl)
